### ISSUE TITLE
Why does content-negotiation result in a 302/303 redirection?

### ISSUE THREAD

**Issue by ThijsBroersen** [schemaorg/schemaorg#2231]
Why does content negotiation (e.g. https://schema.org) result in a 302 redirection? Why not returning the requested representation of the resource, that's what content negotiation is for.. right? 
And why is schema.org using extensions for all resource representations, instead of just returning the requested representation? Again, that's what content negotiation is for.. right? (e.g. 'curl -v -H "Accept: application/ld+json" https://schema.org/Person' results in a 303 to https://schema.org/Person.jsonld)

All these redirections create unnecessary overhead because additional round-trips are required.

This could also result in developers just suffixing .jsonld for all json-ld requests but when the json-ld representation is not available it takes again more round-trips to find other representations. While content negotiation could solve this by a single requests with multiple acceptable content types.

**Comment by RichardWallis** [comment]
@ThijsBroersen 
Thank you for your question.

There are a few reasons for use of the web standard 3** redirect behaviour of the Schema.org serving application.

Requests to http://schema.org/* receive a 301 (Moved Permanently) response redirecting to the _https_ equivalent URL.  The default way of achieving this result as implemented by the cloud environment hosting the service.

Requests for individual term definitions (eg. https://schema.org/Book, https://schema.org/about), default to returning html, but are (303) redirected to an appropriate source of data in the the appropriate format if a supported form is requested in the Accept header.

Requests for format ```application/ld+json``` to the root URL (https://schema.org) are treated as a special case, resulting in a (302) redirect to the JSON-LD Context document describing the whole vocabulary.

As described [in the documentation](https://schema.org/docs/developers.html#conneg), this is currently experimental functionality.  It was implemented, as an enhancement to the initial html only site, using a similar [dereferencing] approach to the W3C [Best Practice Recipes for Publishing RDF Vocabularies](https://www.w3.org/TR/swbp-vocab-pub/#negotiation). 

For site loading and performance reasons each format is prepared and cached separately.  The 3** redirect approach works well in balancing the high volume html request rates we receive with the low volume requirements for other formats.

In answer to your "_that's what content negotiation is for.. right?_" question, the answer is '_sort of_'.  I would phrase it more like '_to supply or direct you to a response in the appropriate format_'.

```curl -v -L -H "Accept: application/ld+json" https://schema.org/Person```

As you point out, it can result in an extra round trip for the non-html output formats. However with edge caching provided in the hosting cloud environment, appropriate use of HEAD requests and the ETag header, it appears to be an acceptable overhead.

It is my understanding is that more often than not the non-html formats are used in development tools etc., where such an overhead would only be experienced occasionally.  Unlike a production environment where the term URLs are used as identifiers and not requests to the Schema.org server.

Maybe you have a need that differs from that understanding, which would be great to hear about.

**Comment by ThijsBroersen** [comment]
@RichardWallis thanks for your elaborate response. I am working on a linked data graph database and use term URLs as identifiers. 

W.r.t. the http -> 301 -> https; within the resources all identifiers still use the plain http-protocol-prefix. 
(When) will these migrate to https? From a programmer perspective I would like to be directed as close to the destination as possible. Graph-instances do not always have a history of cached 30x response, so I has to learn again that resources are moved when building context definitions. 

(A perhaps more general ld question) As the protocol is part of the identifier, are identifiers which differ only in the protocol part assumed to be the same entity?

**Comment by RichardWallis** [comment]
> I am working on a linked data graph database and use term URLs as identifiers.

@ThijsBroersen  Good to hear you are using them as identifiers in your graph (as against some - who our logs seem to tell us - appear to request the URL every time they come across the URL in their code!)

As to the http/https questions..  

In pure data terms, the the two forms of identifier (http://... & https://) are explicitly different.  The assumption that they both identify the same entity is therefore [currently] in the hands of the data consumer.  As an impartial observer, I believe that some of the major consumers of Schema.org formatted data do make this assumption.

Experimentally, and only in the RDFa form of data embedded in the html version of term definitions, the following was added (eg. for the Book term)*: ```<link  property="sameAs" href="https://schema.org/Book"/>```  Establishing an explicit relationship between both identifier URLs.  This, as yet, has not been followed through by adding the same ```sameAs``` triple in all output serialisations.
( * _Note this feature was missing in 3.5, but will return in 3.6_)

I can see the benefits of harmonising the default protocol for the website and the canonical identifier form for terms, both to https.  However, as previously discussed in several forums, it is not that simple, with some 8 years worth of data consumers building applications and datasets with the http based protocol.  Such a move should be well thought through, perhaps starting with the extension of the ```sameAs``` triple into all serialisations.

As to 302/303 redirects in general... Although it is understood how we might do this, I see little general desire to embark on such a core change in operation.

**Comment by ThijsBroersen** [comment]
What do you think of this:
If I do 'curl -v -H "Accept:application/ld+json" https://www.w3.org/2000/01/rdf-schema#isDefinedBy' I get the jsonld-response I expect and a [Content-Location](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Location) header with an alternate location (rdf-schema.jsonld).

**Comment by RichardWallis** [comment]
Looks like another valid but different implementation to that we have used for Schema.org

BTW the  Content-Location header is directing to the file that contains the definition of the whole rdfs vocabulary.

I have labelled this issue as "Closed and Noted".    

Addressing how content negotiation functionality is implemented may be part of future implementation review(s), but as of the moment I see little general desire to embark on such a core change.
